src/Compilers/CSharp/Test/Semantic/Semantics/ExtensionTypeTests.cs

```diff
@@ -2198,41 +2683,286 @@ public void UnderlyingType_MissingTypeSyntax()
+            // explicit extension R3 for C<object?> : R { }
+            Diagnostic(ErrorCode.ERR_UnderlyingTypesMismatch, "R3").WithArguments("R3", "C<object?>", "C<object>").WithLocation(2, 20)
+            );
+
+        var src4 = """
+explicit extension R3 for C<object> : R { }
+""";
+        // PROTOTYPE the nullability warning should be silenced here (oblivious context)
+        var comp4 = CreateCompilation(src4, references: new[] { comp.EmitToImageReference() });
+        comp4.VerifyDiagnostics(
+            // (1,20): error CS9216: Extension 'R3' has underlying type 'C<object>' but a base extension has underlying type 'C<object>'.
+            // explicit extension R3 for C<object> : R { }
+            Diagnostic(ErrorCode.ERR_UnderlyingTypesMismatch, "R3").WithArguments("R3", "C<object>", "C<object>").WithLocation(1, 20)
+            );
+
+        return;
+
+        static void validate(ModuleSymbol module)
+        {
+            var r = module.GlobalNamespace.GetTypeMember("R");
+            VerifyExtension<TypeSymbol>(r);
+            Assert.Equal("C<System.Object!>", r.ExtensionUnderlyingTypeNoUseSiteDiagnostics.ToTestDisplayString(includeNonNullable: true));
+            Assert.Empty(r.BaseExtensionsNoUseSiteDiagnostics);
+        }
     }
 
     [Fact]
     public void UnderlyingType_TupleWithElementNames()
     {
         var src = """
-explicit extension R for (int a, int b) { }
+public explicit extension R for (int a, int b) { }
 """;
         var comp = CreateCompilation(src);
-        comp.VerifyDiagnostics(
+        comp.VerifyDiagnostics();
+
+        CompileAndVerify(comp, symbolValidator: validate, sourceSymbolValidator: validate);
+
+        var comp1 = CreateCompilation(src);
+        comp1.MakeTypeMissing(WellKnownType.System_Runtime_CompilerServices_TupleElementNamesAttribute);
+        comp1.VerifyDiagnostics(
+            // (1,33): error CS8137: Cannot define a class or member that utilizes tuples because the compiler required type 'System.Runtime.CompilerServices.TupleElementNamesAttribute' cannot be found. Are you missing a reference?
+            // public explicit extension R for (int a, int b) { }
+            Diagnostic(ErrorCode.ERR_TupleElementNamesAttributeMissing, "(int a, int b)").WithArguments("System.Runtime.CompilerServices.TupleElementNamesAttribute").WithLocation(1, 33)
+            );
+
+        var src2 = """
+explicit extension R2 for (int a, int b)  : R { }
+""";
+        var comp2 = CreateCompilation(src2, references: new[] { comp.EmitToImageReference() });
+        comp2.VerifyDiagnostics();
+
+        var src3 = """
+explicit extension R3 for (int, int) : R { }
+""";
+        var comp3 = CreateCompilation(src3, references: new[] { comp.EmitToImageReference() });
+        comp3.VerifyDiagnostics(
+            // (1,20): error CS9216: Extension 'R3' has underlying type '(int, int)' but a base extension has underlying type '(int a, int b)'.
```

Comment:
Should this be just a warning?

Reply:
Not at the moment. I'm starting with strict matching requirements, but we'll likely allow some variant. For example, your base extension might extension `object`, but you could extend `string` (narrowing).
I have `TODO2 we need rules to only allow an underlying type that is compatible with the base extensions.` in the spec, but can also add a PROTOTYPE in this test.